File: framework/src/main/scala/com/software/dm/swallow/stormy/scala/algoac/impl/AhoCorasickCharacterTree.scala
package com.software.dm.swallow.stormy.scala.algoac.impl

import java.util

import com.software.dm.swallow.stormy.scala.algoac.inter.AbstractAhoCorasick
import com.software.dm.swallow.stormy.scala.algoac.pojo.ResultSetEntity


/**
  *
  *
  * @author darlingming@126.com
  *         2016
  * @version v1.0.0.1
  *
  */
class AhoCorasickCharacterTree() extends AbstractAhoCorasick[Character](new CharacterState(0)) {
  /**
    *
    */
  override def search(keys: Any): util.Collection[Any] = {
    if (!this.prepared) throw new IllegalStateException("can't start search until prepare()")
    val set = new util.HashSet[Any]
    this.search(keys, set)
    set
  }

  override def addKeyWord(keyword: Any): AhoCorasickCharacterTree = {
    var tmpState = this.rootState
    if (keyword.isInstanceOf[String]) {
      for (c <- keyword.toString.toCharArray) {
        tmpState = tmpState.addState(c)
      }
      tmpState.addOutput(keyword)
    } else if (keyword.isInstanceOf[ResultSetEntity]) {
      val reg = keyword.asInstanceOf[ResultSetEntity]
      for (c <- reg.getKeyWord.toCharArray) {
        if (c >= 65 && c <= 90) {
          val c1 = (32 + c).toChar
          tmpState = tmpState.addState(c1)
        } else {
          tmpState = tmpState.addState(c)
        }

      }
      tmpState.addOutput(reg)
    }
    this
  }

  /**
    *
    * @param keys
    * @param set
    */
  def search(keys: Any, set: util.Set[Any]): Unit = {
    if (!this.prepared) throw new IllegalStateException("can't start search until prepare()")
    var currentState = this.rootState
    for (c <- keys.toString.toCharArray) {
      if (c >= 65 && c <= 90) {
        val c1 = (32 + c).toChar
        currentState = this.getState(c1, currentState)
      } else {
        currentState = this.getState(c, currentState)
      }


      if (null != currentState.getOutput) set.addAll(currentState.getOutput)
    }
  }


}

File: framework/src/main/java/com/software/dm/swallow/stormy/algoac/impl/AhoCorasickCharacterTree.java
package com.software.dm.swallow.stormy.algoac.impl;

import com.software.dm.swallow.stormy.algoac.inter.AbstractAhoCorasick;
import com.software.dm.swallow.stormy.algoac.inter.AbstractState;
import com.software.dm.swallow.stormy.algoac.pojo.ResultSetEntity;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

/**
 * @author DearM
 * @version v1.0.0.1
 * @Description
 * @date 2016
 */
public class AhoCorasickCharacterTree extends AbstractAhoCorasick<Character> {

    public AhoCorasickCharacterTree() {
        super(new CharacterState());
    }

    /**
     *
     */
    public Collection<Object> search(Object keys) {
        if (!this.prepared) {
            throw new IllegalStateException("can't start search until prepare()");
        }
        Set<Object> set = new HashSet<Object>();
        this.search(keys, set);
        return set;
    }

    /**
     *
     */
    public AhoCorasickCharacterTree addKeyWord(Object keyword) {
        AbstractState<Character> tmpState = this.rootState;
        if (keyword instanceof String) {
            for (Character c : keyword.toString().toCharArray()) {
                tmpState = tmpState.addState(c);
            }
            tmpState.addOutput(keyword);
        } else if (keyword instanceof ResultSetEntity) {
            ResultSetEntity reg = (ResultSetEntity) keyword;
            for (Character c : reg.getKeyWord().toCharArray()) {
                if (c >= 65 && c <= 90) {
                    c = (char) (32 + c);
                }
                tmpState = tmpState.addState(c);
            }
            tmpState.addOutput(reg);
        }

        return this;
    }

    /**
     *
     */
    public void search(Object keys, Set<Object> set) {
        if (!this.prepared) {
            throw new IllegalStateException("can't start search until prepare()");
        }
        AbstractState<Character> currentState = this.rootState;
        for (Character c : keys.toString().toCharArray()) {
            if (c >= 65 && c <= 90) {
                c = (char) (32 + c);
            }
            currentState = this.getState(c, currentState);
            if (null != currentState.getOutput()) {
                set.addAll(currentState.getOutput());
            }
        }
    }

}
